xdsl/utils/lexer.py

```diff
@@ -173,8 +172,42 @@ def from_span(cls, span: Span | None) -> StringLiteral | None:
 
     @property
     def string_contents(self):
-        # TODO: is this a hack-job?
-        return ast.literal_eval(self.text)
+        # Common case, so we optimize for it
+        if "\\" not in self.text:
+            return self.text[1:-1]
+        res = list[str]()
+        text = self.text[1:-1]
+        i = 0
+        while i < len(text):
+            c = text[i]
+            i += 1
+            if c != "\\":
+                res.append(c)
+                continue
+            if i >= len(text):
+                raise ValueError(self, "invalid string literal")
+            c1 = text[i]
+            i += 1
+            match c1:
+                case "n":
+                    res.append("\n")
+                case "t":
+                    res.append("\t")
+                case "\\":
+                    res.append("\\")
+                case '"':
+                    res.append('"')
+                case _ if c1 in hexdigits:
+                    if i >= len(text):
+                        raise ValueError(self, "invalid string literal")
+                    c2 = text[i]
+                    i += 1
+                    if c2 not in hexdigits:
+                        raise ValueError(self, "invalid string literal")
+                    res.append(chr(int(c1 + c2, 16)))
```

Comment:
Similarly to above, this does not handle larger unicode codepoints.

I think there might be a better approach than iterating over the whole input one-by-one (makes longer hex escapes easier):

 - use `split(r"\")` to get parts
 - look at the start of each part in `parts[1:]` to identify escape sequences (can be e.g. regex)
 - convert escape sequences and leave the other string parts in tact for printing.

Reply:
the most expensive case is definitely the hex codes, and I have a suspicion that it's not very common if we only hit them now, so it may be worth adding a separate regex to escape `\n` et al, and do another `'\' in res` check to bail early if we don't need to invoke the state machine